msyslog-v1.08a+smac/src/man/peochk.8
.\"	$CoreSDI: peochk.8,v 1.7.2.2.2.1.4.8 2001/11/20 09:56:23 alejo Exp $
.\"
.\" Copyright (c) 2001
.\"	Core-SDI SA. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Core-SDI SA nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 10, 2000
.Dt PEOCHK 8
.Os Core-SDI
.Sh NAME
.Nm peochk
.Nd Initial key generator and integrity log file checker
.Sh SYNOPSIS
.Nm peochk
.Op Fl f Ar logfile
.Op Fl g
.Op Fl h
.Op Fl i Ar key0file
.Op Fl k Ar keyfile
.Op Fl l
.Op Fl m Ar hash_method
.Op Fl q
.Op Ar logfile
.Sh DESCRIPTION
.Nm peochk
generates the initial key file and checks log files generated by
.Xr syslogd 8
using
.Em peo output module
.Xr om_peo 8 .
The options are as follows:
.Bl -tag -width Ds
.It Fl f Ar logfile
Specify the pathname of a log file, if
.Ar logfile
is not specified using this option, data is read from standard input
and the pathname is used only to generate reports and/or to obtain the
key files pathnames when the
.Fl k
and/or
.Fl i
options are not specified; the default is
.Pa /var/log/messages .
.It Fl g
Generates two key files with an initial key into them, one in binary mode
(
.Ar keyfile ,
to be used by
.Em peo output module
) and the other in ascii mode (
.Ar key0file
), the admin should put the
last one into a secure place and remove it from the specified path (see 
.Fl i
and 
.Fl k
options); when this option is not specified
.Nm
is in check mode.
.It Fl h
Displays a little help.
.It Fl i Ar key0file
Specify the initial key pathname; the default is
.Ar keyfile
pathname with a "0" char added at the end (see 
.Fl k
option).
.It Fl k Ar keyfile
Specify the key pathname (this file is used by the
.Em peo output module
to generate a hash key from the last logged message); the default is
.Pa /var/ssyslogd/xxx.key
where 
.Pa xxx
is
.Ar logfile
(specified with 
.Fl f
option or without it) with all '/' replaced by '.'.
.It Fl l
Used only in check mode to detect the first corrupted line; it is ignored
when specified with the
.Fl g
option.
.It Fl m Ar hash_method
Specifies the hash method used to generate the keys,
.Ar hash_method
should be one of 
.Cm md5, sha1,
or
.Cm rmd160;
the default is
.Cm sha1.
.It Fl q
Quiet mode; prints '0' on stdout when logfile is not corrupted, and '1' or
line number (see 
.Fl l
option) when the logfile is corrupted.
.El
.Sh EXAMPLES
If you want to protect the
.Pa /var/log/authlog
file you can:
.Pp
.Bl -enum
.It
run the command:
.Pp
.Dl peochk -g -f /var/log/authlog -i authkey0 -m rmd160
.Pp
this will generate the
.Pa /var/ssylog/var.log.authlog.key
file with the initial key in binary mode and the
.Ar ./authkey0
file with that key translated to ascii, the hash method used to generate
the key is
.Cm rmd160;
you should memorice the contents of
.Ar ./authkey0
file and
.Xr rm 1
it.
.Pp
.It
Edit
.Xr syslog.conf 5
file and enable
.Em peo output module
with something like this:
.Pp
.Dl auth.info	%classic /var/log/authlog %peo -m rmd160 -l -k /var/ssyslog/.var.log.authlog.key
.Pp
.It
Inform new changes on
.Xr syslog.conf 5
to
.Xr syslogd 8 :
.Pp
.Dl kill -HUP `cat /var/run/syslog.pid`
.Pp
.It
When you believe that someone owned your machine you can:
.Pp
.Dl peochk -m rmd160 -f /var/log/authlog -i mykey
.Pp
the contents of
.Ar mykey
should be the same as 
.Ar ./authkey0
generated in step 1; with the command above you can verify that the
file was (or not) corrupted (it is important not to forget the 
.Fl m
option because the default used is 
.Cm sha1
and the keys generated was using 
.Cm rmd160
).
.El
.Sh SEE ALSO
.Xr syslog.conf 5 ,
.Xr om_peo 8 ,
.Xr syslogd 8
.Sh BUGS
Submit bugs at this project's Sourceforge Bug reporting system at:
http://sourceforge.net/tracker/?func=add&group_id=25741&atid=385117
You may also report them directly to the authors; send an email to
core.devel.alat@corest.com, describing the problem the most you can,
containing also machine description, hardware description, the
configuration file (/etc/syslog.conf), the OS description, and the
invoking command line.
The more you describe the bug, the faster we can fix it.


html/ChangeLog.txt
$Id$

Feature Request:

Bugs:
1.  Change code to allow for tailing spaces when looking at the expression field

3/10/2004
	-  Removing any reference to company name

3/9/2004
	-  Changed out company logo
	-  Added GPL notice to all libraries

1/21/2004
	-  Found a bug where a host would email off an alert with no text.  I've put a stop gap fix in.

1/13/2004
        -  Fixed bug with msyslog.pgsql where SET STORAGE didn't have a space before it.
        -  Fixed bug where hour in time of rules would default to 18:00 for no good reason
	-  Fixed two problems with cloning rules:  1)  order was not preserved  2)  an imported version of the database couldn't clone because fields that were empty needed to have non-null defaults applied
	-  Forgot to merge in latest VPN reports

1/6/2004
        -  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/6/2004
	-  Added 'left menu' support for ACID and MRTG groups in security framework should the ever be added 8)

1/1/2004
	-  Fixed spelling error in with the word 'threshold' in the rules schema
	-  Fixed a bug in vacuumdb where it was 'ANALYZ' not 'ANALYZE'

12/19/2003
	-  Updated code to look into /opt/apache

12/10/2003
	-  Updated database to schema to not use compressed text fields(we'll see how this performs)
	-  working on adding interfaces to more lock data and other new stats with PostgreSQL V7.4
	-  cleaned up some button descriptions on the maintenance page, also added a lock view as well as a settings view

12/9/2003
	-  Finished rule.php support for basic timer maintenance.  Need to add another page to graft timers onto rules.
	-  processlogs.php is now setup to support rule timers
	-  need to write rule expiration process 
	-  started converting away from compressed text in hopes of providing faster data retrieval.... also pulled OIDs from DB definition

12/8/2003
	-  Adding support for date and time based rules with date ranges, day of week selections, as well as deleted rules, need to add interface to control rule timestamp properties

12/7/2003
	-  finished adding accumulation thresholds

12/3/2003
	-  adding support for both types of thresholds.  Need to update web pages to reflect new radio buttons.
	-  added web configuration support for supression thresholds and accumulating thresholds, now onto updating the log processor
	-  basic supression works(kinda), need to verify functionality

12/2/2003
	-  customer profiles can now have multiple hosts added at once.
	-  fixed a stupid bug where <prev|refresh|next> didn't behave right.  next would stop working if you started at oct-26-2003 and it would stay on oct-26-2003
	-  Added web-based framework and database schema to support alert supression thresholds

12/1/2003
	-  Updated processlogs.php code to better deal with single entry/no rules vs mutliple rules

11/19/2003
	-  Updated mail table to enforce unique login ids(effectively stopping two processes from running at the same time, one will crash and die(safely))
	-  Updated openmail and closeopenmail to use transaction support since PostgreSQL no longer does server-side auto-commit(ie. convert everything to transactions)
	-  Transaction support should now be officially added, will do some testing

11/18/2003
	-  Updating program for support with PostGreSQL V7.4
	-  Fixed host process table to reflect the fact that multiple hosts are in there by default	

10/13/2003
	-  Added alert total to bottom of alert aggregation

10/6/2003
	-  finished support for alert aggregation
	-  updated 1stview to pull the current time and date
	-  fixed host.php bug where you could expire syslogs but not be forced to expire syslogs

10/5/2003
	-  still adding support for alert aggregation.  Basic aggregation works plus alert zooming but need to add support for across the board for other alert queries

10/3/2003
	-  added ability to un/suspend log processors from the web interface
	-  updated maintenance to rebuild all indexes in an better manner(ie. grab the index list from the DB rather then by hand)
	-  added additional framework to do alert aggregation interface.  Need to add 'aggregation code' for display

9/24/2003
	-  weeklyindexrebuild.php now pulls all indexes from the system and rebuilds them.  The result is that the system will now rebuild any new indexes without manual reconfiguration.

9/15/2003
	-  launchid was not initiated correctly in the clonerule.

9/2/2003
	-  Fixed a bug where '\'s at the end of a line caused problems because we were not properly dealing with them in general.  Fixed that.

8/23/2003	
	-  Created another bug when fixing 5000 line paging.  Timestamp was thrown off in view.php

8/18/2003	
	-  Syncing changes from production smt environment:  vacuumtsyslog.php
	-  Updated weeklyindexrebuild.php to account for the correct indexes
	-  Updated maint.php to account for the three new indexes for the launch program section

8/13/2003
	-  view.php has had several updates.  Paging should now be fixed.  Multiple searches appeared not to be working correctly.
	-  Needed to add lastid as hidden var if the variable was set
	-  Needed to use urlencode on top of htmlspecialcharacters, filters were broken because of it

8/11/2003
	-  1stfilter.php doesn't list 'global' filters that you down own
	-  modified view.php to not let the user save a filter with no description
	-  added support to delete all of a user's filters(ie. do before delete)
	-  another problem popped up with filter.php when I added the delete user filter option

8/1/2003	
	-  processlogs.php now supports launching external programs!

7/31/2003
	-  Updated vacuumdb.php to do a full vacuum of the TSyslog table.  Why?  Because the system doesn't reclaim disk space or use old delete space for some reason
	-  Almost finished adding launch program code, need to test.

7/27/2003
	-  Added weeklyindexrebuild.php which rebuilds all indexes at 5am Sunday morning

7/23/2003

msyslog-v1.08a+smac/src/peo/peochk.c
/*	$CoreSDI: peochk.c,v 1.35.2.2.4.9 2001/11/20 09:56:24 alejo Exp $	*/

/*
 * Copyright (c) 2001, Core SDI S.A., Argentina
 * All rights reserved
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither name of the Core SDI S.A. nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * peochk - syslog -- Initial key generator and integrity log file checker
 *
 * Author: Claudio Castiglia, Core-SDI SA
 *
 *
 * peochk [-f logfile] [-g] [-i key0file] [-k keyfile] [-l]
 *        [-m hash_method] [-q] [logfile]
 *
 * supported hash_method values:
 *			md5
 *			rmd160
 *			sha1
 *
 * defaults:
 *	logfile: 	/var/log/messages
 *	keyfile:	/var/ssyslog/.var.log.messages.key
 *	hash_method:	sha1
 *
 * NOTES:
 *	1) When logfile is specified without the -f switch, the data is
 *	   read from the standard input 
 *	2) If logfile is specified using both -f switch and without it,
 *	   the -f argument is used and data is read from that file
 *	3) If logfile is specified but not the keyfile, this will be
 *	   /var/ssyslog/xxx.key where xxx is the logfile with all '/'
 *	   replaced by '.'
 *	4) If -l switch is specified, peochk detects the line number
 *	   corrupted on logfile
 *	5) -q means quiet mode
 *
 */

#include "config.h"

#include <sys/param.h>
#include <sys/stat.h>
#include <sys/syslog.h>
#include <sys/types.h>
#include <sys/uio.h>

#include <errno.h>
#include <fcntl.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#if TIME_WITH_SYS_TIME
# include <sys/time.h>
# include <time.h>
#else
# if HAVE_SYS_TIME_H
#  include <sys/time.h>
# else
#  include <time.h>
# endif
#endif
#include <unistd.h>
#include <netdb.h>

#include "hash.h"
#include "../modules.h"
#include "../syslogd.h"

#define CHECK	0x01
#define QUIET	0x02
#define ST_IN	0x04

int	 actionf;
char	*corrupted = "corrupted";
char	*default_logfile = "/var/log/messages";
char	*keyfile;
char	*key0file;
char	*logfile;
char	*macfile;
int	 method;

extern char *optarg;


/*
 * release allocated memory
 */
void
release(void)
{
	if (keyfile != default_keyfile)
		free(keyfile);
	if (key0file)
		free(key0file);
	if (logfile != default_logfile)
		free(logfile);
	if (macfile)
		free(macfile);
}


/*
 * usage
 */
void
usage(void)
{
	fprintf (stderr,
		"Usage:\n"
		"Check mode:\n"
		"    peochk [-h] [-l] [-f logfile] [-i key0file] [-k keyfile]\n"
		"           [-m hash_method] [-q] [logfile]\n\n"
		"Initial key generator mode:\n"
		"    peochk -g [-h] [-k keyfile] [-m hash_method] [logfile]"
		"\n\n");
	exit(-1);
}


/*
 * readline()
 */
int
readline(int fd, char *buf, size_t len, FILE **f)
{
	char	*st;
	size_t	 i;

	if (*f == NULL) {
		*f = fdopen(fd, "r");
		if (*f == NULL)
			return (-1);
	}
	st = fgets(buf, len, *f);
	if (st != NULL) {
		i = strlen(buf) - 1;
		if (*(buf + i) == '\n')
			*(buf + i ) = '\0';
		return (i);
	}
	*buf = '\0';
	return (0);
}


/*
 * eexit()
 *	Prints a message on stdout and exits with a status
 */
void
eexit(int status, char *fmt, ...)
{
	va_list ap;
	if (fmt) {
		va_start(ap, fmt);
#ifdef HAVE_VPRINTF
		vfprintf(stdout, fmt, ap);
#elif defined(HAVE_DOPRNT)
		_doprnt(stdout, fmt, ap);
#else
#error No vfprintf and no doprnt
#endif
		va_end(ap);
	}
	exit(status);
}
	
	
/*
 * check: read logfile and check it
 */
void
check(void)
{
	FILE *finput;
	int   i;
	int   input;
	int   mfd;
	unsigned char  key[41];
	int   keylen;
	unsigned char  lastkey[21];
	int   lastkeylen;
	int   line;
	unsigned char  mkey1[21];
	int   mkey1len;
	unsigned char  mkey2[21];
	int   mkey2len;
	char  msg[MAXLINE];
	int   msglen;

	/* open logfile */
	if (actionf & ST_IN)
		input = STDIN_FILENO;
	else if ( (input = open(logfile, O_RDONLY, 0)) == -1) {
		perror(logfile);
		exit(-1);
	}

	mfd = 0;	/* shutup gcc */

	/* open macfile */
	if (macfile)
		if ( (mfd = open(macfile, O_RDONLY, 0)) == -1) {
			perror(macfile);
			exit(-1);
		}

	/* read initial key (as ascii string) and tranlate it to binary */
	if ( (i = open(key0file, O_RDONLY, 0)) == -1) {
		perror(key0file);
		exit(-1);
	}
	if ( (keylen = read(i, key, 40)) == -1) {
		perror(key0file);
		exit(-1);
	}
	if (!keylen) {
		if (actionf & QUIET)
			eexit(1, "1\n");
		else
			eexit(1, "(1) %s: %s\n", key0file, corrupted);
	}
	key[keylen] = 0;
	asc2bin(key, key);
	keylen >>= 1;
	close(i);

	/* read last key */
	if ( (i = open(keyfile, O_RDONLY, 0)) == -1) {
		perror(keyfile);
		exit(-1);
	}
	if ( (lastkeylen = read(i, lastkey, 20)) == -1) {
		perror(keyfile);
		exit(-1);
	}
	if (!lastkeylen) {
		if (actionf & QUIET)
			eexit(1, "1\n");
		else
			eexit(1, "(1) %s: %s\n", keyfile, corrupted);
	}
	close(i);

	/* test both key lenghts */
	if (lastkeylen != keylen) {
		if (actionf & QUIET)
			eexit(1, "1\n");
		else
			eexit(1, "(1) %s and/or %s %s\n", key0file, keyfile,
			    corrupted);
	}

	/* check it */
	line = 1;
	finput = NULL;
	while ( (msglen = readline(input, msg, MAXLINE, &finput)) > 0) {
		if (macfile) {
			if ( ((mkey1len = mac2(key, keylen,
			    (unsigned char *) msg, msglen, mkey1)) < 0) ||
			    ((mkey2len = read(mfd, mkey2,
			    mkey1len)) < 0) ) {
				perror(macfile);
				exit(-1);
			}
			if ((mkey2len != mkey1len) || memcmp(mkey2, mkey1,
			    mkey1len)) {
				if (actionf & QUIET)
					eexit(1, "%i\n", line);
				else
					eexit(1, "(%i) %s %s on line %i\n",
					    line, logfile, corrupted, line);
			}
			line++;
		}
		if ( (keylen = mac(method, key, keylen,
		    (unsigned char *) msg, msglen, key)) == -1) {
			perror("fatal");
			exit(-1);
		}
	}

	if (finput != NULL)
		fclose(finput);

	if (macfile != NULL)
		close(mfd);

	if (i < 0) {
		fprintf(stderr, "error reading logs form %s : %s\n",
		    (actionf & ST_IN) ? "standard input" : logfile,
		    strerror(errno));
		exit(-1);
	}

	if (memcmp(lastkey, key, keylen)) {
		if (actionf & QUIET) 
			eexit(1, "1\n");
		else
			eexit(1, "(1) %s %s\n", logfile, corrupted);
	}
	if (actionf & QUIET)
		eexit(0, "0\n");
	else
		eexit(0, "(0) %s file is ok\n", logfile);
}


/*
 * generate:
 *	generate initial key and write it on keyfile and key0file
 *	in the last file data is written as ascii string
 */
void
generate(void)
{
	int	 kfd;
	int	 k0fd;
	unsigned char	 key[20];
	unsigned char	 keyasc[41];
	int	 keylen;
	unsigned char	 randvalue[20];

	if (getrandom(randvalue, 20) < 0) {
		release();
		perror("getrandom");
		exit(-1);
	}
	if ( (keylen = mac(method, NULL, 0, randvalue, 20, key)) == -1) {
		release();
		perror("fatal");
		exit(-1);
	}
	if ( (kfd = open(keyfile, O_WRONLY|O_CREAT|O_EXCL,
	    S_IRUSR|S_IWUSR)) == -1) {
		release();
		perror(keyfile);
		exit(-1);
	}
	if ( (k0fd = open(key0file, O_WRONLY|O_CREAT|O_EXCL,
	    S_IRUSR|S_IWUSR)) == -1) {
		unlink(keyfile);
		close(kfd);
		release();
		perror(key0file);
		exit(-1);
	}

	/* write key 0 */
	write(kfd, key, keylen);
	write(k0fd, bin2asc(keyasc, key, keylen), keylen << 1);
	close(kfd);
	close(k0fd);
}
	

/*
 * main
 */
int
main(int argc, char **argv)
{
	int	 ch;
	int	 mac;

	/* integrity check mode, stdin */
	actionf = CHECK | ST_IN;

	/* default values */
	logfile = default_logfile;
	keyfile = default_keyfile;
	key0file = NULL;
	mac = 0;
	macfile = NULL;
	method = SHA1;

	/* parse command line */
	while ( (ch = getopt(argc, argv, "f:ghi:k:lm:q")) != -1) {
		switch (ch) {
			case 'f':
				/* log file (intrusion detection mode) */
				if (logfile != default_logfile)
					free(logfile);
				if ( (logfile = strrealpath(optarg)) == NULL) {
					release();
					perror(optarg);
					exit(-1);
				}
				actionf &= ~ST_IN;
				break;
			case 'g':
				/* generate new keyfile and initial key */
				actionf &= ~CHECK;
				break;
			case 'i':
				/* key 0 file */
				if (key0file)
					free(key0file);
				if ( (key0file = strdup(optarg)) == NULL) {
					release();
					perror(optarg);
					exit(-1);
				}
				break;
			case 'k':
				/* keyfile */
				if (keyfile != default_keyfile)
					free(keyfile);
				if ( (keyfile = strdup(optarg)) == NULL) {
					release();
					perror(optarg);
					exit(-1);
				}
				break;
			case 'l':
				mac = 1;
				break;
			case 'm':
				/* hash method */
				if ( (method = gethash(optarg)) < 0) {
					release();
					usage();
				}
				break;
			case 'q':
				/* quiet mode */
				actionf |= QUIET;
				break;
			case 'h':
			default:
				release();
				usage();
		}

	}

	/* check logfile specified without -f switch */
	argc -= optind;
	argv += optind;
	if (argc && (actionf & ST_IN))
		if ( (logfile = strrealpath(argv[argc-1])) == NULL) {
			release();
			perror(argv[argc-1]);
			exit(-1);
		}

	/* if keyfile was not specified converted logfile is used instead */
	if (keyfile == default_keyfile && logfile != default_logfile) {
		char *tmp;

		if ( (tmp = strallocat("/var/ssyslog/", logfile)) == NULL) {
			release();
			perror("buffer for keyfile");
			exit(-1);
		}
		strdot(tmp+13);
		if ( (keyfile = strallocat(tmp, ".key")) == NULL) {
			free(tmp);
			release();
			perror("buffer for keyfile");
			exit(-1);
		}
		free(tmp);
	}

	/* if key0file was not specified create one */
	if (key0file == NULL)
		if ( (key0file = strkey0(keyfile)) == NULL) {
		release();
		perror("creating key0 file");
		exit(-1);
		}

	/* create macfile */
	if (mac)
		if ( (macfile = strmac(keyfile)) == NULL) {
			release();
			perror("creating mac file");
			exit(-1);
		}

	/* execute action */
	if (actionf & CHECK)
		check();
	else
		generate();

	release();
	return (0);
}

